scripts/build.py

```diff
@@ -174,47 +184,113 @@ def _ensure_files_exist(filepaths):
 
     Raises:
-        Exception: Raised if one or more of the files does not exist.
+        OSError: One or more of the files does not exist.
     """
     for filepath in filepaths:
         if not os.path.isfile(filepath):
-            raise Exception('File %s does not exist.' % filepath)
+            raise OSError('File %s does not exist.' % filepath)
+
+
+def get_file_count(directory_path):
+    """Count total number of file directory, subtracting ignored files.
+
+    Args:
+        directory_path: str. Directory to be walked.
 
+    Returns:
+        int. Total number of files minus ignored files.
+    """
+    total_file_count = 0
+    for _, _, files in os.walk(directory_path):
+        total_file_count += len(files)
+        for filename in files:
+            # Ignore files with certain extensions.
+            if any(filename.endswith(p)
+                   for p in FILE_EXTENSIONS_TO_IGNORE):
+                total_file_count -= 1
+    if directory_path == ASSETS_DEV_DIR:
+        # Add 1 to account for the added hashes.js.
+        total_file_count += 1
+    return total_file_count
 
-def process_html(source_path, target_path, file_hashes):
+
+def _compare_file_count(source_path, target_path):
+    """Ensure that two dir's file counts match.
+
+    Args:
+       source_path: str. Source of staging files.
+       target_path: str. Final filepath or built file's directory.
+
+    Raises:
+        ValueError: Raised if both dir do not have the same file count.
+    """
+    source_dir_file_count = get_file_count(source_path)
+    target_dir_file_count = get_file_count(target_path)
+    if source_dir_file_count != target_dir_file_count:
+        print 'Comparing %s vs %s' % (source_path, target_path)
+        raise ValueError(
+            '%s files in source dir != %s files in target dir.' % (
+                source_dir_file_count, target_dir_file_count))
+
+
+def _match_filename_with_hashes(filename, file_hashes):
+    """Ensure that filepath is hashed correctly.
+
+    Args:
+        filename: str. Filepath to be matched.
+        file_hashes: dict(str, str). Dictionary of file hashes.
+
+    Raises:
+        ValueError: Raised if hash dict is empty.
+        ValueError: Raised if filename does not contain hash.
+        KeyError: Raised if filename's hash does not match hash dict entries.
+    """
+    # Final filepath example: base.240933e7564bd72a4dde42ee23260c5f.html.
+    if not file_hashes:
+        raise ValueError('Hash dict is empty')
+    file_hash = re.findall(r"([a-fA-F\d]{32})", filename)
+    # Convert current /build path to /backend_prod_files path.
+    if not file_hash:
+        raise ValueError('%s is expected to contain hash' % filename)
+    if file_hash[0] not in file_hashes.values():
+        raise KeyError(
+            'Hashed file %s does not match hash dict keys' % filename)
+
+
+def process_html(source_path, minified_html_file_stream, file_hashes):
```

Comment:
Perhaps minified_html_file_stream --> target_file_stream, for parity.

You might want a source_file_stream as well, for tests. That will allow you to keep the tests independent of reading the file system. (The file system may change in the future and these tests are important for ensuring that the build system is still correct regardless of that.)

Reply:
Yep, using 2 file stream object for input and output for _process_html and adjusted its unit test,